File: app/src/main/java/com/smart/novel/test/ACT_Login.java
package com.smart.novel.test;

import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.RadioButton;

import com.smart.framework.library.base.BaseActivity;
import com.smart.framework.library.bean.ErrorBean;
import com.smart.framework.library.common.utils.AppSharedPreferences;
import com.smart.framework.library.common.utils.CommonUtils;
import com.smart.framework.library.common.utils.TimeCount;
import com.smart.framework.library.net.ReqCallBack;
import com.smart.framework.library.net.RequestManager;
import com.smart.framework.library.netstatus.NetUtils;
import com.smart.novel.MyApplication;
import com.smart.novel.R;
import com.smart.novel.bean.UserEntity;
import com.smart.novel.db.bean.TestBean;
import com.smart.novel.db.gen.TestBeanDao;
import com.smart.novel.db.manager.DbManager;
import com.smart.novel.global.API;
import com.smart.novel.util.SharePreConstants;

import org.greenrobot.greendao.Property;

import java.util.ArrayList;
import java.util.HashMap;

import butterknife.Bind;
import butterknife.OnClick;

/**
 * Created by JoJo on 2018/6/21.
 * wechat：18510829974
 * description：
 */
public class ACT_Login extends BaseActivity {
    @Bind(R.id.rb_get_code)
    RadioButton rbGetCode;
    @Bind(R.id.et_phone)
    EditText etPhone;
    @Bind(R.id.et_password)
    EditText etPassword;
    @Bind(R.id.ll_login)
    LinearLayout llLogin;
    //验证码倒计时
    private TimeCount timer;
    //倒计时间隔时间
    private final int countDownInterval = 1000;
    //验证码倒计时总时间
    private final int msgTime = 60 * 1000;
    private AppSharedPreferences sharePre;
    private TestBean testBean;

    @Override
    protected int getContentViewLayoutID() {
        return R.layout.act_login;
    }

    @Override
    protected void handleStatusBar() {

    }

    @Override
    protected void getBundleExtras(Bundle extras) {

    }

    @Override
    protected void initViewsAndEvents() {
        timer = new TimeCount(msgTime, countDownInterval, rbGetCode, mContext);
        sharePre = new AppSharedPreferences(this);

        String token = sharePre.getString(SharePreConstants.USER_SIGN);
        if (!TextUtils.isEmpty(token)) {
            finish();
        } else {
            etPhone.setText(sharePre.getString(SharePreConstants.USER_NAME));
            etPhone.setSelection(etPhone.getText().length());
        }
    }


    @OnClick({R.id.rb_get_code, R.id.rb_login})
    public void onClick(View view) {
        switch (view.getId()) {
            //获取验证码
            case R.id.rb_get_code:
                if (getLoginUser() != null) {
                    CommonUtils.makeEventToast(mContext, getLoginUser().getMsg(), false);
                }
                break;
            //登录
            case R.id.rb_login:
                if (doVerifyEmpty()) {
                    requestLogin();
                    showDialogLoading("正在登入...");
                }
                break;
        }
    }

    /**
     * 校验用户输入
     *
     * @return
     */
    private boolean doVerifyEmpty() {
        if (TextUtils.isEmpty(etPhone.getText().toString())) {
            CommonUtils.makeEventToast(MyApplication.getInstance(), "账号不能为空", false);
            return false;
        }
        if (TextUtils.isEmpty(etPassword.getText().toString())) {
            CommonUtils.makeEventToast(MyApplication.getInstance(), "密码不能为空", false);
            return false;
        }
        return true;
    }

    private void requestLogin() {
        HashMap<String, String> params = new HashMap<>();
        params.put("a", "login");
        params.put("name", etPhone.getText().toString());
        params.put("password", etPassword.getText().toString());
        RequestManager.getInstance().requestGetByAsyn(API.SERVER_IP, params, new ReqCallBack<UserEntity>() {
            @Override
            public void onReqSuccess(UserEntity userEntity) {
                dismissDialogLoad();
                CommonUtils.makeEventToast(MyApplication.getInstance(), "登录成功", false);
                //存储用户信息
                sharePre.putString(SharePreConstants.USER_SIGN, userEntity.getSign() == null ? "" : userEntity.getSign());
                sharePre.putString(SharePreConstants.USER_ID, userEntity.getId() + "");
                sharePre.putString(SharePreConstants.USER_NAME, userEntity.getName());
                sharePre.putString(SharePreConstants.TYPE, userEntity.getType());
                sharePre.putString(SharePreConstants.BRANCH_NAME, userEntity.getBranchName());
                sharePre.putBoolean(SharePreConstants.LOGOUT, false);
//                Constants.USER_ID = sharePre.getString(SharePreConstants.USER_ID);
//                Constants.USER_SIGN = sharePre.getString(SharePreConstants.USER_SIGN);
//                Constants.TYPE = sharePre.getString(SharePreConstants.TYPE);
//                Constants.BRANCH_NAME = sharePre.getString(SharePreConstants.BRANCH_NAME);
                finish();
            }

            @Override
            public void onFailure(String result) {
                dismissDialogLoad();
                CommonUtils.makeEventToast(MyApplication.getInstance(), result, false);
            }

            @Override
            public void onReqFailed(ErrorBean error) {
                dismissDialogLoad();
                testBean = new TestBean();
                testBean.setMsg(error.getMsg());
                DbManager.getInstance().insertOrReplace(TestBean.class, testBean);
                CommonUtils.makeEventToast(MyApplication.getInstance(), error.getMsg(), false);
            }
        });
    }

    /**
     * 测试GreenDao
     *
     * @return
     */
    public TestBean getLoginUser() {
        ArrayList<Property> properties = new ArrayList<Property>();
        properties.add(TestBeanDao.Properties.Msg);
        // 匹配条件
        ArrayList<Object> objects = new ArrayList<Object>();
        objects.add(testBean.getMsg());
        TestBean userInfo = null;
        try {
            userInfo = (TestBean) DbManager.getInstance().query(TestBean.class, properties, objects);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return userInfo;
    }

    @Override
    public void showBusinessError(ErrorBean error) {

    }

    @Override
    protected long getRefreshTime() {
        return 0;
    }

    @Override
    protected boolean isApplyKitKatTranslucency() {
        return false;
    }

    @Override
    protected View getLoadingTargetView() {
        return null;
    }


    @Override
    protected void onNetworkConnected(NetUtils.NetType type) {

    }

    @Override
    protected void onNetworkDisConnected() {

    }

    @Override
    protected boolean isApplyStatusBarTranslucency() {
        return false;
    }

    @Override
    protected boolean isBindEventBusHere() {
        return false;
    }

    @Override
    protected boolean toggleOverridePendingTransition() {
        return false;
    }

    @Override
    protected TransitionMode getOverridePendingTransitionMode() {
        return null;
    }
}

File: app/src/main/java/com/smart/novel/ui/ACT_Login.kt
package com.smart.novel.ui

import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.View
import butterknife.OnClick
import com.smart.framework.library.bean.ErrorBean
import com.smart.framework.library.common.log.Elog
import com.smart.framework.library.common.utils.AppSharedPreferences
import com.smart.framework.library.common.utils.TimeCount
import com.smart.framework.library.common.utils.login.VerifyInputUtils
import com.smart.novel.MyApplication
import com.smart.novel.R
import com.smart.novel.base.BaseMVPActivity
import com.smart.novel.bean.UserBean
import com.smart.novel.db.manager.DbManager
import com.smart.novel.mvp.contract.LoginContract
import com.smart.novel.mvp.model.LoginModel
import com.smart.novel.mvp.presenter.LoginPresenter
import com.smart.novel.util.BroadCastConstant
import com.smart.novel.util.SharePreConstants
import kotlinx.android.synthetic.main.act_login.*

/**
 * Created by JoJo on 2018/8/23.
 * wechat:18510829974
 * description: 登录页面
 */
class ACT_Login : BaseMVPActivity<LoginPresenter, LoginModel>(), LoginContract.View {

    var countDownTmer: TimeCount? = null
    //验证码倒计时时间：60s
    val msgTime = 60 * 1000
    //倒计时间隔时间
    val intervalTime = 1000
    var sharePre: AppSharedPreferences? = null
    override fun getContentViewLayoutID(): Int {
        return R.layout.act_login
    }

    override fun getBundleExtras(extras: Bundle?) {

    }

    override fun startEvents() {
        sharePre = AppSharedPreferences(MyApplication.context)
        //验证码倒计时
        countDownTmer = TimeCount(rb_get_code, this)

        // addTextChangedListener ???不好使
//        RxTextView.textChanges(et_phone_number).subscribe { text ->
//            Elog.e("","textChanges")
//        }

//        val etPhoneObservable = RxTextView.textChanges(et_phone_number)
//        val etCodeObservable = RxTextView.textChanges(et_code)

//        // 相当于合并
//        Observable.combineLatest(etPhoneObservable, etCodeObservable,
//                BiFunction<CharSequence, CharSequence, Any> { phone, code ->
//                    // 设置按钮是否可用(或者改变背景颜色)
//                    val b = !TextUtils.isEmpty(phone) && !TextUtils.isEmpty(code)
//                    rb_login.isChecked = b
//                    return@BiFunction this
//                })
        initListener()
    }

    private fun initListener() {
        et_phone_number.addTextChangedListener(object : TextWatcher {

            override fun afterTextChanged(s: Editable?) {
            }

            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                Elog.e("txr", s.toString())
                if (!TextUtils.isEmpty(s.toString())) {
                    rb_get_code.setTextColor(ContextCompat.getColor(MyApplication.context, R.color.color_3AC270))
                } else {
                    rb_get_code.setTextColor(ContextCompat.getColor(MyApplication.context, R.color.color_999999))
                }
            }
        })
    }

    @OnClick(R.id.rb_login, R.id.rb_get_code)
    fun onClick(view: View) {
        when (view.id) {
            R.id.rb_get_code -> {
                if (VerifyInputUtils.doValidatePhone(mContext, et_phone_number)) mMvpPresenter.sendCode(et_phone_number.text.toString())
            }
            R.id.rb_login -> {
                if (VerifyInputUtils.doValidatePhoneAndCode(mContext, et_phone_number, et_code)) mMvpPresenter.login(et_phone_number.text.toString(), et_code.text.toString())

            }
        }
    }

    override fun isBindEventBusHere(): Boolean {
        return false
    }

    override fun showBusinessError(error: ErrorBean?) {

    }

    override fun showException(error: ErrorBean?) {
    }

    override fun sendCode(data: Any) {
        countDownTmer!!.start()
    }

    override fun login(userInfo: UserBean) {
        //保存登录状态及用户信息
        MyApplication.isLogin = true
        sharePre!!.putString(SharePreConstants.USER_ID, userInfo.id.toString())
        sharePre!!.putString(SharePreConstants.USER_PHONE, userInfo.phone)
        sharePre!!.putString(SharePreConstants.USER_NAME, userInfo.user_name)
        sharePre!!.putString(SharePreConstants.USER_HEAD_AVATAR, userInfo.chat_head_uri)
        sharePre!!.putBoolean(SharePreConstants.LOGOUT, false)
        DbManager.getInstance().insertOrReplace(UserBean::class.java, userInfo)
        sendBroadcast(BroadCastConstant.LOGIN_SUCCESS)
        finish()
    }

}